### ISSUE TITLE
Investigate local autocorrect algorithms 

### ISSUE THREAD

**Issue by dessalines** [dessalines/thumb-key#44]
A few for java:

- https://developer.android.com/develop/ui/views/touch-and-input/spell-checker-framework
- https://github.com/C-FWES/AutoCorrect
- https://github.com/TeodorDyakov/autocorrect
- https://github.com/GameGenesis/java-autocomplete

**Comment by Zireael07** [comment]
Speaking of, I find that it autocorrects i to I, which is annoying when I happen NOT to be writing in English...

**Comment by dessalines** [comment]
Currently that's part of auto-capitalize, but I agree, it should be removed as its correct only for english.

**Comment by DeflateAwning** [comment]
This would be great!

**Comment by racuna** [comment]
IIRC MultilingO stores couples of words locally and then a number of occurrences to build a pseudo-smart word prediction: e.g. "I want a beer". 

I:want:1
want:a:1
a:beer1

"I want to walk"

i:want:2
want:to:1
to:walk

So, the keyboard could suggest "want" after the user writes "I", then, if the user accepts: "want", the keyboard could suggest "a" or "to" because they both have the same "weight" of 1.

**Comment by racuna** [comment]
And can be backed up in plain text

**Comment by racuna** [comment]
yeah, of course. I meant about having it backed up locally and "editable"

**Comment by pwd-github** [comment]
I wanted to revive this conversation by introsucing a repository. https://github.com/fengjian0106/Minuum-Fleksy-Fuzzy-Matching

Could this be helpfull? I think, maybe so

**Comment by Geobert** [comment]
That’s the reason I gave up on MessagEase in the first place. With open source alternative, my hope of using this kind of keyboard is back alive :D

**Comment by breversa** [comment]
Agreed! :-)
Not having autocorrect/autocompletion is the number 1 reason why I haven’t switched full-time to Thumk-Key. :-|

**Comment by Hate9** [comment]
@pwd-github probably not? I'm pretty sure the main benefits of what minuum does rely pretty heavily on a keyboard where the only input is *tapping*. *could* be relevant though, and it's certainly an interesting resource.

**Comment by Hate9** [comment]
ooo, an "on-the-side" option would be perfect, yeah!

**Comment by seguri** [comment]
hey everyone!

I've just finished adding autocompletion to [my fork](https://github.com/seguri/thumb-key/tree/autocomplete). It's more of a PoC rather than a full product, but it shows what can be done :) I've spent no time tweaking the graphical appearance of the buttons. Right now I'm using a custom dictionary defined with some Italian words. An example:


![image](https://github.com/dessalines/thumb-key/assets/2405157/f2048adb-d8a6-4511-b607-0263359118c4)

**Comment by dessalines** [comment]
Nice!

It'd be a good idea to show the suggestions above the keyboard, like a lot of the other ones do. 

I did some searching around, and unfortunately it doesn't look like android gives access to anything but the user dictionary: https://developer.android.com/reference/android/provider/UserDictionary.html

So we might have to think about the best ways to either import or bundle dictionaries into thumb-key.

**Comment by Hate9** [comment]
I think on-the-side makes a lot of sense for a lot of use-cases, but I think we should probably have it be a settings toggle. For a lot of phones, you won't be using that space anyway so it's a great option, but for (for instance) [mine](https://www.unihertz.com/products/jelly-2), that space is smaller or otherwise needed (to say nothing of the two-hands layouts).

**Comment by Hex5DA** [comment]
some ideas i had a while back, maybe a bit overreaching for this discussion. if they're irrelevant lmk.
- the middle button in the 3x3 (or the spacebar?) autocompletes the next predicted word
- the angle which a swipe needs to land in to type a character grows/shrinks if that character is more/less likely to be the next character
- the list of contender words for autocompletion is shown to the side of the keyboard (not above)

**Comment by racuna** [comment]
> So we might have to think about the best ways to either import or bundle dictionaries into thumb-key.

Hi, there's a OSS project with dictionaries: GNU Aspell

https://ftp.gnu.org/gnu/aspell/dict/0index.html

**Comment by dessalines** [comment]
Once @seguri has their PR ready to look at, we can try to integrate that in some way.

**Comment by seguri** [comment]
What features should my fork integrate to consider it a sufficiently valid PR?
The purpose of my fork was primarily to tackle something I had never done before: passing data around with idiomatic Android data structures and building a simple autocomplete engine. I consider it complete for developers who can take my fork and write their own dictionary to get autocompletion.
I fear that getting my code PR-ready might require lots of time (80% of the time to complete the last 20% of the work, as per the [pareto principle](https://stackoverflow.com/questions/608748/how-to-avoid-the-80-20-rule-in-software-development)).

**Comment by dessalines** [comment]
Start the PR and mark it as draft and we can go over what else it needs.

**Comment by jm355** [comment]
One thing that might be good is a gesture to select the suggestion. If three suggestions get displayed, you could use the top row of gestures in the top left square, where the top left corner selects the leftmost suggestion, top middle selects the middle suggestion, and top right corner selects rightmost suggestion

**Comment by Atemu** [comment]
Perhaps gestures could be used to select different forms of the same word such as tenses or cases. E.g. typing `brea` would offer "break" and swiping e.g. right would produce "breaking", up "broke" etc.

**Comment by Eletroboss** [comment]
Hello,
I've wanted autocorrect in Thumb Key for a long time (it's the main reason I don't use it)
I made a branch and added it myself at https://github.com/Eletroboss/thumb-key/.
That branch is definitely not prepared to be merged, since it doesn't have an "off" button and I'm sure there are a few issues with the composingText, and also it only works for English, and also also the frequency dictionary I found doesn't appear to have a license - but it does work.
Thought it might be useful.

**Comment by dessalines** [comment]
Feel free to open a PR when you've got it in a ready state so we can take a look.

**Comment by racuna** [comment]
Hi again, how about using a 3rd party library instead of reinventing the wheel? e.g. https://github.com/mthebaud/predict4all

**Comment by dessalines** [comment]
Nice! I don't have time to integrate that as I'm busy with other things, but it'd be great if someone would give it a try.

**Comment by jrcarter** [comment]
I like Thumb-Key a lot, but the lack of word suggestions is the reason I mostly continue to use MessageEase (14.2). As I use a full-width keyboard, I'd prefer the suggestions to be across the top, and selected by tapping the desired word.

**Comment by jm355** [comment]
I know the state of this ticket is basically pr or gtfo (but in a polite way 🙂), but just want to link to https://github.com/GrapheneOS/os-issue-tracker/issues/5694 which is somewhat related, and maybe thumbkey could fix it with a good solution? I'm not sure how android separates system vs user spell checkers.

**Comment by dessalines** [comment]
Android spell-checking seems to be mostly deprecated on most android variants. I checked two devices and they have it hidden now.
